### PR TITLE
refactor(web): move websocket to crux

### PR DESCRIPTION
**Pull request by m8vago**
Move the ws server to crux.
Implement ws adapter.
Implement custom decorators.
Send subscription messages on the socket instead of http.

### CODE REVIEW

File: web/crux/src/app/node/node.message.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import { IsDate } from 'class-validator'
+import { NodeConnectionStatus } from '../shared/shared.dto'
+
+export const WS_TYPE_NODE_EVENT = 'event'
+export class NodeEventMessage {
+  id: string
+
+  status: NodeConnectionStatus
+
+  address?: string
```
Comment (polaroi8d) [inline web/crux/src/app/node/node.message.ts:11]: Can we need to add `@isOptional()` here?

File: web/crux/src/main.ts [hunk 2]
```diff
@@ -77,6 +83,10 @@ const bootstrap = async () => {
   const apiOptions = loadGrpcOptions('api', configService.get<string>('GRPC_API_PORT'))
   const httpOptions = configService.get<string>('HTTP_API_PORT', '1848')
 
+  app.useGlobalGuards(app.get(JwtAuthGuard))
```
Comment (polaroi8d) [inline web/crux/src/main.ts:86]: If we are using globbally, we should delete from controllers.

File: web/crux/src/domain/utils.ts [hunk 3]
```diff
@@ -61,6 +61,17 @@ export const toTimestamp = (date: Date): Timestamp => {
   return { seconds, nanos }
 }
 
+export const fromTimestamp = (timestamp: Timestamp): Date => {
```
Comment (polaroi8d) [inline web/crux/src/domain/utils.ts:64]: We don't have this mapper already somewhere else?

File: web/crux/src/domain/agent.ts [hunk 4]
```diff
@@ -184,7 +188,7 @@ export class Agent {
 
     this.eventChannel.next({
       id: this.id,
-      status: NodeConnectionStatus.UNREACHABLE,
+      status: 'unreachable',
```
Comment (polaroi8d) [inline web/crux/src/domain/agent.ts:191]: Using enums, way better thans strings, IMHO

File: web/crux/src/domain/agent.ts [hunk 5]
```diff
@@ -357,3 +361,13 @@ export type AgentToken = {
   iss: string
   iat: number
 }
+
+export type AgentEvent = {
+  id: string
+  status: NodeConnectionStatus
+  address?: string
+  version?: string
```
Comment (polaroi8d) [inline web/crux/src/domain/agent.ts:369]: There are use cases, where we don't have agent versions? cc. @nandor-magyar

File: web/crux/src/app/node/node.message.ts (hunk 1 above)
Comment (m8vago) [reply web/crux/src/app/node/node.message.ts:11]: I am not quite sure yet if the validator runs on the ws messages, probably we'll have to tweak it a little, but yeah we could add it.

File: web/crux/src/main.ts (hunk 2 above)
Comment (m8vago) [reply web/crux/src/main.ts:86]: yeah that's the idea.

File: web/crux/src/domain/utils.ts (hunk 3 above)
Comment (m8vago) [reply web/crux/src/domain/utils.ts:64]: didn't found it. we may have it in the ui code.

File: web/crux/src/domain/agent.ts (hunk 4 above)
Comment (m8vago) [reply web/crux/src/domain/agent.ts:191]: These are typed string literals. The enum was a generated by grpc.

File: web/crux/src/domain/agent.ts (hunk 5 above)
Comment (m8vago) [reply web/crux/src/domain/agent.ts:369]: Not really,  but if the agent isn't connected we can't send its address, hence it's optional.

File: web/crux/src/app/token/jwt-auth.guard.ts [hunk 6]
```diff
@@ -39,34 +54,60 @@ export default class JwtAuthGuard extends AuthGuard('jwt') {
...
+        req.identity = await this.kratos.getIdentityById(userId)
       } catch {
         return false
       }
     }
 
     return activated
   }
+
+  private canActivateWs(context: ExecutionContext): boolean {
```
[the remark this answers is not in the document]
Comment (m8vago) [reply web/crux/src/app/token/jwt-auth.guard.ts:66]: `canActivate` is an interface method of guards, so to match that terminology we better off with `canActivateWs` imho.

File: web/crux/src/app/node/node.ws.gateway.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+  @WsAuthorize()
+  async onAuthorize(@TeamId() teamId: string, @IdentityFromSocket() identity: Identity): Promise<boolean> {
+    return await this.teamService.checkUserActiveTeam(teamId, identity)
+  }
+
+  @WsSubscribe()
+  onSubscribe(@TeamId() teamId: string): Observable<WsMessage> {
+    return from(this.service.subscribeToNodeEvents(teamId)).pipe(
+      mergeAll(),
```
Comment (nandor-magyar) [inline web/crux/src/app/node/node.ws.gateway.ts:34]: Is this 100% necessary?

**Review by nandor-magyar — COMMENTED** [review]
Partially reviewed.

File: web/crux/src/app/registry/registry-api/v2-api-client.ts [hunk 8]
```diff
@@ -31,7 +31,9 @@ class RegistryV2ApiClient implements RegistryApiClient {
     const res = await this.fetch('/')
     if (!res.ok) {
       const errorMessage = `Version request failed with status: ${res.status} ${res.statusText}`
-      throw res.status === 401 ? unauthorizedError(errorMessage) : internalError(errorMessage)
+      throw res.status === 401
+        ? new UnauthorizedException(errorMessage)
+        : new InternalServerErrorException(errorMessage)
```
Comment (polaroi8d) [inline web/crux/src/app/registry/registry-api/v2-api-client.ts:36]: Can only these two errors occur?

File: web/crux/src/app/registry/registry.http.controller.ts [hunk 9]
```diff
@@ -20,14 +10,10 @@ import {
   ApiTags,
 } from '@nestjs/swagger'
 import { Identity } from '@ory/kratos-client'
-import HttpLoggerInterceptor from 'src/interceptors/http.logger.interceptor'
-import PrismaErrorInterceptor from 'src/interceptors/prisma-error-interceptor'
-import { API_CREATED_LOCATION_HEADERS } from 'src/shared/const'
-import UuidValidationGuard from 'src/guards/uuid-params.validation.guard'
 import UuidParams from 'src/decorators/api-params.decorator'
+import { API_CREATED_LOCATION_HEADERS } from 'src/shared/const'
```
Comment (polaroi8d) [inline web/crux/src/app/registry/registry.http.controller.ts:14]: Where is the usage of this `API_CREATED_LOCATION_HEADERS`?

File: web/crux/src/app/registry/registry.http.controller.ts [hunk 10]
```diff
@@ -43,14 +29,7 @@ const ROUTE_REGISTRY_ID = ':registryId'
...
 @Controller(ROUTE_REGISTRIES)
 @ApiTags(ROUTE_REGISTRIES)
-@UsePipes(
-  new ValidationPipe({
-    // TODO(@robot9706): Move to global pipes after removing gRPC
-    transform: true,
-  }),
-)
-@UseInterceptors(HttpLoggerInterceptor, PrismaErrorInterceptor, CreatedWithLocationInterceptor)
-@UseGuards(JwtAuthGuard, UuidValidationGuard, RegistryTeamAccessGuard)
```
Comment (polaroi8d) [inline web/crux/src/app/registry/registry.http.controller.ts:53]: Why removed `Guards and Interceptors`?

File: web/crux/src/app/registry/registry-api/v2-api-client.ts (hunk 8 above)
Comment (polaroi8d) [reply web/crux/src/app/registry/registry-api/v2-api-client.ts:36]: Resolved.

File: web/crux/src/app/registry/registry.http.controller.ts (hunk 10 above)
Comment (polaroi8d) [reply web/crux/src/app/registry/registry.http.controller.ts:53]: Resolved.

File: web/crux/src/app/node/node.ws.gateway.ts (hunk 7 above)
Comment (m8vago) [reply web/crux/src/app/node/node.ws.gateway.ts:34]: yep

**Pull request merged**